src/codegen/llvm/codegen_llvm_visitor.cpp

```diff
@@ -62,458 +60,352 @@ static bool can_vectorise(const ast::CodegenForStatement& statement, symtab::Sym
     return collected.empty();
 }
 
-llvm::Value* CodegenLLVMVisitor::create_gep(const std::string& name, llvm::Value* index) {
-    llvm::Type* index_type = llvm::Type::getInt64Ty(*context);
-    std::vector<llvm::Value*> indices;
-    indices.push_back(llvm::ConstantInt::get(index_type, 0));
-    indices.push_back(index);
-
-    return ir_builder.CreateInBoundsGEP(lookup(name), indices);
-}
-
-llvm::Value* CodegenLLVMVisitor::codegen_indexed_name(const ast::IndexedName& node) {
-    llvm::Value* index = get_array_index(node);
-    return create_gep(node.get_node_name(), index);
+llvm::Value* CodegenLLVMVisitor::accept_and_get(const std::shared_ptr<ast::Node>& node) {
+    node->accept(*this);
+    return ir_builder.pop_last_value();
 }
 
-llvm::Value* CodegenLLVMVisitor::codegen_instance_var(const ast::CodegenInstanceVar& node) {
-    const auto& member_node = node.get_member_var();
-    const auto& instance_name = node.get_instance_var()->get_node_name();
-    const auto& member_name = member_node->get_node_name();
-
-    if (!instance_var_helper.is_an_instance_variable(member_name))
-        throw std::runtime_error("Error: " + member_name + " is not a member of the instance!");
+void CodegenLLVMVisitor::create_external_function_call(const std::string& name,
+                                                       const ast::ExpressionVector& arguments) {
+    if (name == "printf") {
+        create_printf_call(arguments);
+        return;
```

Comment:
just to clarify : `printf` works only for serial execution, right?
(don't need to fix it here, it's used for debugging purpose only)

Reply:
Yes, `printf` does not check that input types are coherent with the provided string (e.g. "%d", etc.). So it works well for scalar code, but if vectorized it prints some random things - I think vector is treated as an int or something similar.

It would be nice to fix it however, and should be rather easy actually, just replicate printf calls for every element.

Example:
```c++
printf("%d", i) // i is vector
```
would yield
```c++
<i1, i2, i3, i4>  // < can be use or some other bracket-like symbol
```

The caveat here is that the format string needs to be broken apart, for additional printf calls, etc. (I don't think you can printf with llvm vector)

Reply:
Actually, what we can do is to define an extern C interface that prints llvm vectors: and replace/insert these printing function if needed:

Something like:
```c++
extern "C" void print_some_type(SomeType<int8_t> *T) {
  std::cout << "<";
  // Deal with type here
  std::cout << ">";
}
```

Reply:
Yes, thats one option. There is something similar has been done in ISPCP https://github.com/ispc/ispc/. I will try to find out what's been done.

Anyway, this is future enhancement and we don't have to worry about it now. I will mark it resolve.